### DISCUSSION TITLE
Golang API calls to PB as admin

### DISCUSSION THREAD

**Discussion by normanmadrid** [pocketbase/pocketbase#3020] [Q&A] [answered]
Greetings good sirs! I have been desperately trying to get the code snippet here: https://github.com/pocketbase/pocketbase/issues/378 working, the one written by r--w. Im authenticating successfully but getting a response message 'Only admins can perform this action', when I try to list a collection. I am indeed authenticated as a user admin, but it's not seeing that for some reason. Here is my code: https://gist.github.com/normanmadrid/a8596195463b2e925df7a89ef31c82dc 
please help.

**Comment by ganigeorgiev** [comment, marked as answer]
I think the issue in your case is the use of `Admin` token prefix which is no longer needed as of [PocketBase v0.8+](https://github.com/pocketbase/pocketbase/blob/master/CHANGELOG.md#api).

In other words, try replacing https://gist.github.com/normanmadrid/a8596195463b2e925df7a89ef31c82dc#file-gistfile1-txt-L224 (and anywhere else):

```go
old: c.client.SetHeader("Authorization", "Admin "+auth.Token)

new: c.client.SetHeader("Authorization", auth.Token)
```

**Reply by normanmadrid** [reply]
Hot damn! It worked you are a life saver gani!

**Reply by normanmadrid** [reply]
Sorry but I tried using the getList as a getFullList (apparently its limited to 500), how would one accomplish that without an SDK?

**Reply by ganigeorgiev** [reply]
Yes, the max `perPage` limit is 500. There is no option to return in a single HTTP request all items from a collection as it could be very inefficient with large collections.

`getFullList` under the hood sends batched `getList` requests (either concurrently or one by one) until there are no more items (_with v0.17.0 you can also set the `?skipTotal=1` query parameter to speed up the queries since the total counters will be skipped_). This could be more efficient compared to 1 big request because the GC will not have to allocate excessive amount of memory all at once and the db connection will be busy for a shorter time allowing the processing of other operations in between the different requests.

In general, if your collection has more than 1k+ records, I'd advice against fetching everything at once and instead try to make use of the existing pagination or implement some sort of lazy loading.

**Reply by normanmadrid** [reply]
Ya you see the scenario is like this, I'm processing data to be inserted into pocketbase, but I don't want to process items in that data that have previously been inserted into pocketbase, so I'm trying to compare IDs in that data with existing IDs in pocketbase, I'm expecting the items in this particular collection to be reaching 500k or so. Maybe I should approach this problem from another angle.

**Reply by ganigeorgiev** [reply]
You can try to do the processing on batches and concurrently (if not already). Personally I think 500k records are a lot to be fetched at once with a single HTTP request and alternative approach should be researched.

It would depend on your specific use case, but it may be better to do the processing on lower level instead of using the web APIs, aka. to interact directly with the PocketBase app instance and db abstraction via Go (_or with the JS app hooks if you  are using v0.17+_).

**Reply by normanmadrid** [reply]
Yup was thinking the same, thanks so much for the rapid response and helpful insight!

**Reply by normanmadrid** [reply]
I'm trying to find an example of an API create rule where it will look up a list of IDs to see if the ID of the record being created already exists in the collection, and if it does exist in the list it won't, but I'm coming up empty. Is it possible to implement a rule like that? or would I need to create this logic in some createRecordHook?

**Reply by ganigeorgiev** [reply]
I'm not sure I follow.

If the id you are trying to check is from the same collection as the record that is being created then **you don't have to do anything in addition** and you can leave that to the built in collection validators (aka. you can't have 2 records with the same id in the same collection).

If the id you are trying to check if from some other collection, then I guess you can use the `@collection.*` rule, eg. `@collection.someOtherCollection.id != id` (_roughly translates to "require every id of someOtherCollection to be different than the current one"_).

But querying the database directly will be more efficient than the above. Could you elaborate?

**Reply by normanmadrid** [reply]
I have a field 'transactionId' that holds the original id of the object im recreating /inserting into pocketbase into my collection, before i insert i want to make sure that no record holds the same 'transactionId', which means that data object has been processed and already inserted in the past. I figure maybe ill just populate the Id field of the record instead of using 'transactionId' as i create these records from the data im processing like you said, I tried using the 'unique constraint' option for transactionId but its still letting me insert new records with the same transactionId.

![image](https://github.com/pocketbase/pocketbase/assets/71151590/724cc142-4ac5-44a5-88c4-e6e80bae80fd)

**Reply by ganigeorgiev** [reply]
> I tried using the 'unique constraint' option for transactionId but its still letting me insert new records with the same transactionId.

Make sure that the index is actually "unique", aka. the index expression starts with something like "CREATE **UNIQUE** INDEX ....".

_I've just tested it locally and it is working fine for me. We also have integration and unit tests to ensure the correct behavior (eg. see [record_crud_test.go](https://github.com/pocketbase/pocketbase/blob/b3f09ff045cfb915aea8a8334d0f4f17d236952d/apis/record_crud_test.go#L1200))_

**Reply by normanmadrid** [reply]
ya this slipped by me, thanks again gani!
